src/regexp.js

```diff
@@ -69,87 +109,303 @@ const fuzzPrintableAscii = f => {
+    // invalid braced quantifier intentionally never generated ... it is always an error
+  ];
+  if (state.unicode) {
+    characterFuzzers.push(fuzzDecimalEscape);
   }
-);
+  return choose(...characterFuzzers)(state);
+}
+
+const fuzzClassAtomDash = f => choose(f => '-', fuzzClassAtom)(f);
 
+const fuzzClassAtom = state => {
+  if (state.tooDeep()) return '';
+  state = state.goDeeper();
+  let characterFuzzers = [
+    fuzzPrintableAsciiExcept(['\\', ']', '-']),
+    f => `\\${oneOf(...syntaxCharacters)(f)}`,
+    fuzzCharacterEscapes,
+  ];
+  if (state.unicode) {
+    characterFuzzers.push(
+      f => '\\-'
+    );
+  } else {
+    characterFuzzers.push(
+      f => `\\c${oneOf(...decimal, '_')(f)}`
+    );
+  }
+  return choose(...characterFuzzers)(state);
+};
 
 const fuzzCharacterClass = f => {
   if (f.tooDeep()) return '[]';
   f = f.goDeeper();
-  let source = many(choose(fuzzCharacterClassCharacter, fuzzCharacterClassRange))(f).join('');
-  source = source.replace(/((^|[^\\])(\\\\)*)\\$/g, '$1\\a'); // character class cannot end in an odd number of backslashes
-  return `[${oneOf('^', '-', '')(f)}${source}${oneOf('-', '')(f)}]`;
+  f.globalState.noNumericLookahead = false;
+  let canDash = true;
+  let dashRangeFuzzer = fuzzCharacterClassRange(true);
+  let noDashRangeFuzzer = fuzzCharacterClassRange(false);
+  let canCaret = false;
+  let fuzzToken = f => {
+    let lastWasRange = false;
+    let value = guardEscapedNumeric(f, guardValue(choose(
+      f => {
+        lastWasRange = false;
+        return canDash ? fuzzClassAtomDash(f) : fuzzClassAtom(f)
+      },
+      f => {
+        lastWasRange = true;
+        return canDash ? dashRangeFuzzer(f) : noDashRangeFuzzer(f)
+      }
+    ), value => !canCaret && value.startsWith('^')))(f);
+    canDash = lastWasRange;
+    canCaret = true;
+    return value;
+  };
+  let source = many(fuzzToken)(f).join('');
+  f.globalState.noNumericLookahead = false;
+  return `[${oneOf('^', '')(f)}${source}${oneOf('-', '')(f)}]`;
 };
 
-const fuzzCharacterClassCharacter = f => {
-  let ch;
-  do {
-    ch = choose(fuzzCharacter, oneOf('[', '(', ')', '{', '?', '*', '+', '|', '$'))(f);
-  } while (ch === '-' || ch === ']');
-  return ch;
-};
+const illegalRangeItems = ['\\c'];
+const illegalRangeItemsUnicode = ['\\d', '\\D', '\\w', '\\W', '\\s', '\\S', '\\c'];
 
-const fuzzCharacterClassRange = f => {
-  let a = fuzzCharacterClassCharacter(f);
-  let b = fuzzCharacterClassCharacter(f);
-  if (charVal(b) < charVal(a)) [a, b] = [b, a];
+const fuzzCharacterClassRange = canDash => f => {
+  let toCheckItems = illegalRangeItems;
+  if (f.unicode) {
+    toCheckItems = illegalRangeItemsUnicode;
+  }
+  let a = guardValue(fuzzClassAtom, value => value.length == 0 || toCheckItems.filter(item => value.startsWith(item)).length > 0)(f);
+  let b = guardValue(fuzzClassAtom, value => value.length == 0 || toCheckItems.filter(item => value.startsWith(item)).length > 0)(f);
+  let valueA = charVal(a);
```

Comment:
I think we can get rid of `charVal` if we change the class atom fuzzer to return a code point and then render it when consuming. I really don't like `charVal`.

Reply:
Another option would be to return both the code point and the rendered character simultaneously. This would avoid having to refuzz the output format, while avoiding charVal.